Reference architecture GenAI - final review (#5510)
elastic/docs-content: deploy-manage/reference-architectures/genai-search-high-availability.md

Elastic has performance-tested hardware profiles across the major cloud providers to identify the optimal balance for each node type. Significantly deviating from these tested ratios may appear to reduce costs, but typically leads to degraded performance, query latency spikes, or search scalability.

Dedicated ML nodes are needed when {{infer}} is performed within the cluster, such as running [ELSER](/explore-analyze/machine-learning/nlp/ml-nlp-elser.md) or [custom transformer models](/explore-analyze/machine-learning/nlp/ml-nlp-model-ref.md) locally. These nodes should be dedicated and provisioned with sufficient memory to load models into RAM. When {{infer}} is offloaded to an LLM or embedding model external to the cluster (for example, [Elastic {{infer-cap}} Service](kibana://reference/connectors-kibana/elastic-managed-llm.md), Azure OpenAI, Anthropic, or Bedrock), dedicated ML nodes are not required.

| Type | {{aws}} | Azure | GCP | Physical |
| :---- | :---- | :---- | :---- | :---- |

> | hot | c6gd | f32sv2 | N2 | 16-32 vCPU, 64-256 GB RAM, 2-6 TB NVMe SSD |

Author: @john-wagster , @bradquarry , I've updated the section to solve this. Let me know your thoughts. I think the key is:

> For GenAI search workloads in {{ech}}, use **Vector Search Optimized** profiles as the primary reference, and consider **CPU Optimized** profiles for workloads with higher CPU and disk requirements.

And then link to these 3 docs that I didn't know we had :) 
- [Selecting the right configuration for you on {{aws}}](https://www.elastic.co/docs/reference/cloud/cloud-hosted/ec_selecting_the_right_configuration_for_you)
- [Selecting the right configuration for you on Azure](https://www.elastic.co/docs/reference/cloud/cloud-hosted/ec-azure-configuration-choose)
- [Selecting the right configuration for you on GCP](https://www.elastic.co/docs/reference/cloud/cloud-hosted/ec-gcp-configuration-choose)

Because of that, and because we suggest in the official AWS doc `r6gd` for Vector Search I've added `r6gd` and `c8gd` as the current suggestion (depending on the user needs). And with a final sentence such as:

> These recommendations provide a practical baseline, but available instance families evolve over time as newer provider hardware becomes available. For additional guidance on selecting {{ecloud}} hardware profiles for specific workloads, refer to:
(and the 3 links shared earlier).

---
applies_to:
  deployment:
    self: ga all
    ess: ga all
    ece: ga all
    eck: ga all
products:
  - id: cloud-enterprise
  - id: cloud-kubernetes
  - id: cloud-hosted
---

# GenAI Search - High Availability

This reference architecture illustrates a production-grade, highly available GenAI search solution built on {{es}}. It shows the physical deployment model, logical integration points, and key best practices for implementing a retrieval layer that grounds generative AI responses.

{{es}} can combine [lexical search](/solutions/search/full-text.md), [dense vector search](/solutions/search/vector/dense-vector.md), [sparse vector search](/solutions/search/vector/sparse-vector.md), temporal and geospatial filtering, and hybrid ranking techniques. These capabilities form the foundation for [Retrieval Augmented Generation (RAG)](/solutions/search/rag.md), [agentic workflows](/explore-analyze/ai-features/elastic-agent-builder.md), and AI-assisted applications.

> 

Reviewer: I miss some reader orientation here. The doc jumps quickly into detail.
Would it be possible to add a short section after the intro? Something like:

What you’ll learn:
* Deployment model
* Key components
* Tradeoffs (latency vs cost vs recall)

Or something similar that helps the reader to see what they will learn about.

## Recommended hardware specifications

The following recommendations assist with clusters that are self-deployed on-prem or self-deployed in a cloud provider. With {{ech}}, you can deploy clusters in {{aws}}, Azure, or Google Cloud. Available hardware types and configurations vary across providers, but each offers instance families that meet the performance needs of search and generative AI applications. 

For details, refer to our documentation on [{{ech}} hardware](cloud://reference/cloud-hosted/hardware.md) for {{aws}}, Azure, and GCP. The "Physical" column in the table below provides approximate per-node guidance when self-deploying {{es}} in your own data center, based on equivalent CPU, RAM, and storage profiles.

Elastic has performance-tested hardware profiles across the major cloud providers to identify the optimal balance for each node type. Significantly deviating from these tested ratios may appear to reduce costs, but typically leads to degraded performance, query latency spikes, or search scalability.

> Dedicated ML nodes are needed when {{infer}} is performed within the cluster, such as running [ELSER](/explore-analyze/machine-learning/nlp/ml-nlp-elser.md) or [custom transformer models](/explore-analyze/machine-learning/nlp/ml-nlp-model-ref.md) locally. These nodes should be dedicated and provisioned with sufficient memory to load models into RAM. When {{infer}} is offloaded to an LLM or embedding model external to the cluster (for example, [Elastic {{infer-cap}} Service](kibana://reference/connectors-kibana/elastic-managed-llm.md), Azure OpenAI, Anthropic, or Bedrock), dedicated ML nodes are not required.

Reviewer 2: ```suggestion
Dedicated ML nodes are needed when {{infer}} is performed within the cluster, such as running [ELSER](/explore-analyze/machine-learning/nlp/ml-nlp-elser.md) or [custom transformer models](/explore-analyze/machine-learning/nlp/ml-nlp-model-ref.md) locally. These nodes should provisioned with sufficient memory to load models into RAM. When {{infer}} is offloaded to an LLM or embedding model external to the cluster (for example, [Elastic {{infer-cap}} Service](kibana://reference/connectors-kibana/elastic-managed-llm.md), Azure OpenAI, Anthropic, or Bedrock), dedicated ML nodes are not required.
```
(just avoiding the redundancy that "dedicated nodes should be dedicated")

| master | `c8gd` | `fsv2` | `N2` | 4 vCPU, 16 GB RAM, 256 GB SSD |
| kibana | `c8gd` | `fsv2` | `N2` | 8-16 vCPU, 8-16 GB RAM, 256 GB SSD |

These recommendations provide a practical baseline, but available instance families evolve over time as newer provider hardware becomes available. For additional guidance on selecting {{ecloud}} hardware profiles for specific workloads, refer to:

- [Selecting the right configuration for you on {{aws}}](cloud://reference/cloud-hosted/ec_selecting_the_right_configuration_for_you.md)
- [Selecting the right configuration for you on Azure](cloud://reference/cloud-hosted/ec-azure-configuration-choose.md)
- [Selecting the right configuration for you on GCP](cloud://reference/cloud-hosted/ec-gcp-configuration-choose.md)

For the full list of currently available instance configurations, refer to:

- [{{aws}} default instances](cloud://reference/cloud-hosted/aws-default.md)
- [Azure default instances](cloud://reference/cloud-hosted/azure-default.md)
- [GCP default instances](cloud://reference/cloud-hosted/gcp-default-provider.md)

## Vector search considerations

### Storage and memory

> Approximate nearest-neighbor (ANN) algorithms are dominated by irregular, latency-bound memory access rather than arithmetic. As a result, all vector distance computations occur in off-heap RAM and CPU. To make large-scale in-memory vector search feasible, {{es}} supports [many quantization techniques](elasticsearch://reference/elasticsearch/mapping-reference/dense-vector.md#dense-vector-quantization) for up to a 32x reduction in vector footprint in RAM, and defaults to HNSW with [Better Binary Quantization (BBQ)](https://www.elastic.co/search-labs/blog/better-binary-quantization-lucene-elasticsearch). BBQ is Elastic’s memory-efficient approach for maximizing recall with a low vector memory footprint.

Reviewer 3: "rather than arithmetic" ? what does that mean?

Reviewer: Can we say then something like this?
```suggestion
Approximate nearest-neighbor (ANN) algorithms are dominated by irregular, latency-bound memory access patterns, meaning performance is typically limited by cache misses and data locality rather than raw computation. As a result, all vector distance computations occur in off-heap RAM and CPU. To make large-scale in-memory vector search feasible, {{es}} supports [many quantization techniques](elasticsearch://reference/elasticsearch/mapping-reference/dense-vector.md#dense-vector-quantization) for up to a 32x reduction in vector footprint in RAM, and defaults to HNSW with [Better Binary Quantization (BBQ)](https://www.elastic.co/search-labs/blog/better-binary-quantization-lucene-elasticsearch). BBQ is Elastic’s memory-efficient approach for maximizing recall with a low vector memory footprint.
```

### Vector sizing

- To learn about sizing formulas refer to [Tune approximate kNN search](/deploy-manage/production-guidance/optimize-performance/approximate-knn-search.md#_ensure_data_nodes_have_enough_memory). Use the formulas to calculate the disk needs. 
- For HNSW, the same amount of RAM will be required as the vector index storage. 
- For DiskBBQ, you can provision as little as 5% of disk storage as RAM, as DiskBBQ swaps to disk effectively, but query performance will improve significantly with more data you can fit into RAM.

### Generating vector embeddings

- If generating embeddings on Elastic ML nodes, capacity can be [autoscaled](/deploy-manage/autoscaling.md) using {{ech}} (ECH) or {{eck}} on-prem.
- Elastic provides [world class vector models](/explore-analyze/machine-learning/nlp/ml-nlp-jina.md) from Jina AI that can be used for NLP embeddings, and a re-ranker that can be either self-hosted on ML nodes or used through [Elastic {{infer-cap}} Service](/explore-analyze/elastic-inference/eis.md).

> - Embeddings can also be generated with [Elastic’s {{infer}} API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-inference-put) from many external model providers or internally or externally hosted foundation models.

Reviewer 2: ```suggestion
- Embeddings can also be generated with [Elastic’s {{infer}} API](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-inference-put) from many external model providers, or from internally or externally hosted foundation models.
```
